wled00/FX_fcn.cpp

```diff
@@ -1198,19 +1350,44 @@ void WS2812FX::service() {
         // The blending will largely depend on the effect behaviour since actual output (LEDs) may be
         // overwritten by later effect. To enable seamless blending for every effect, additional LED buffer
         // would need to be allocated for each effect and then blended together for each pixel.
-        [[maybe_unused]] uint8_t tmpMode = seg.currentMode();  // this will return old mode while in transition
-        delay = (*_mode[seg.mode])();         // run new/current mode
 #ifndef WLED_DISABLE_MODE_BLEND
-        if (modeBlending && seg.mode != tmpMode) {
+        [[maybe_unused]] uint8_t tmpMode = seg.currentMode();  // this will return old mode while in transition
+        if (modeBlending && seg.isInTransition()) {
+          // Run new mode first
+          Segment::renderToBuffer(seg.buffer1);
+          delay = (*_mode[seg.mode])();       // render mode into buffer
+
+          // Run old mode into buffer
           Segment::tmpsegd_t _tmpSegData;
           Segment::modeBlend(true);           // set semaphore
           seg.swapSegenv(_tmpSegData);        // temporarily store new mode state (and swap it with transitional state)
           _virtualSegmentLength = seg.virtualLength(); // update SEGLEN (mapping may have changed)
-          uint16_t d2 = (*_mode[tmpMode])();  // run old mode
+          _colors_t[0] = seg.currentColor(0); // update colors to allow for color transitions
+          _colors_t[1] = seg.currentColor(1);
+          _colors_t[2] = seg.currentColor(2);
+          seg.currentPalette(_currentPalette, seg.palette);
+          for (int c = 0; c < NUM_COLORS; c++) _colors_t[c] = gamma32(_colors_t[c]);
+
+          Segment::renderToBuffer(seg.buffer2);
+          uint16_t d2 = (*_mode[tmpMode])();  // render old mode into buffer
+
+          // Clean up
           seg.restoreSegenv(_tmpSegData);     // restore mode state (will also update transitional state)
           delay = MIN(delay,d2);              // use shortest delay
           Segment::modeBlend(false);          // unset semaphore
+          Segment::renderToBuffer(0);
+
+          // Mix everything together
```

Comment:
It would be cleaner to just use one function and do `if (is2D()) {...}` in it.

Reply:
If you feel strongly about it, happy to combine. As of right now, it'd be a pretty long function. As I'm adding more transition styles, that'll only get worse. I'm thinking of ways to split up the implementation of the transitions, perhaps the same way as it's done with effects, where there's one function per effect. That way, we'd also get rid of the switch statement.

Reply:
No, I do not have *strong feelings*. 😁 It just seemed cleaner.